python/beeai_framework/adapters/mcp/serve/server.py

```diff
@@ -0,0 +1,106 @@
+from typing_extensions import TypeVar
+
+from beeai_framework.tools.tool import AnyTool, Tool
+from beeai_framework.tools.types import ToolOutput
+
+try:
+    import mcp.server.fastmcp.prompts as mcp_prompts
+    import mcp.server.fastmcp.resources as mcp_resources
+    import mcp.server.fastmcp.server as mcp_server
+except ModuleNotFoundError as e:
+    raise ModuleNotFoundError(
+        "Optional module [mcp] not found.\nRun 'pip install \"beeai-framework[mcp]\"' to install."
+    ) from e
+
+
+from beeai_framework.serve.server import Server
+from beeai_framework.utils import ModelLike
+from beeai_framework.utils.models import to_model
+
+TInput = TypeVar("TInput", bound=Any, default=Any)
+
+
+class McpServerConfig(BaseModel):
+    """Configuration for the McpServer."""
+
+    transport: Literal["stdio", "sse"] = "stdio"
+    name: str = "MCP Server"
+    instructions: str | None = None
+    settings: mcp_server.Settings[Any] = Field(default_factory=lambda: mcp_server.Settings())
+
+
+class McpServer(
+    Generic[TInput],
+    Server[
+        TInput,
+        mcp_prompts.Prompt | mcp_resources.Resource | Callable[[Any], ToolOutput],
+        McpServerConfig,
+    ],
+):
+    def __init__(self, *, config: ModelLike[McpServerConfig] | None = None) -> None:
+        super().__init__(config=to_model(McpServerConfig, config or McpServerConfig()))
+        self._server = mcp_server.FastMCP(
+            self._config.name,
+            self._config.instructions,
+            **self._config.settings.model_dump(),
+        )
+
+    def serve(self) -> None:
+        for member in self.members:
+            factory = McpServer.get_factory(member)
+            input = factory(member)
+            if callable(input):
+                self._server.add_tool(fn=input, name=member.name, description=member.description)
+            elif isinstance(input, mcp_prompts.Prompt):
+                self._server.add_prompt(input)
+            elif isinstance(input, mcp_resources.Resource):
+                self._server.add_resource(input)
+            else:
+                raise ValueError(f"Input type {type(member)} is not supported by this server.")
+
+        self._server.run(transport=self._config.transport)
+
+    @classmethod
+    def get_factory(
+        cls, member: Any
+    ) -> Callable[
+        [TInput],
+        mcp_prompts.Prompt | mcp_resources.Resource | Callable[[Any], ToolOutput],
+    ]:
+        factory = cls._factories.get(type(member))
+        if factory is None and isinstance(member, Tool):
+
+            def _tool_factory(
+                tool: AnyTool,
+            ) -> Callable[[dict[str, Any]], Coroutine[Any, Any, ToolOutput]]:
+                async def run(input: dict[str, Any]) -> ToolOutput:
+                    result: ToolOutput = await tool.run(input)
+                    return result
+
+                return run
+
+            factory = _tool_factory  # type: ignore
+        assert factory is not None
```

Comment:
You must throw explicitly

`assert` expressions can be disabled in a given environment